packages/rest/src/router/trie-router.ts

```diff
@@ -0,0 +1,53 @@
+// Copyright IBM Corp. 2017, 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {RouteEntry, createResolvedRoute} from './routing-table';
+import {Trie} from './trie';
+import {Request} from '../types';
+import {inspect} from 'util';
+import {BaseRouter} from './router-base';
+
+const debug = require('debug')('loopback:rest:router:trie');
+
+/**
+ * Router implementation based on trie
+ */
+export class TrieRouter extends BaseRouter {
+  private trie = new Trie<RouteEntry>();
+
+  add(route: RouteEntry) {
+    if (super.add(route)) return true;
+    // Add the route to the trie
+    const key = this.getKeyForRoute(route);
+    this.trie.create(key, route);
+    return true;
+  }
+
+  find(request: Request) {
+    const method = request.method.toLowerCase();
+    const path = `/${method}${request.path}`;
+
+    let route = this.simpleRoutes[path];
+    if (route) return createResolvedRoute(route, {});
+
+    const found = this.trie.match(path);
+    debug('Route matched: %j', found);
+
+    if (found) {
+      route = found.node.value!;
```

Comment:
Can you fix the type definition of `trie.match` to communicate the fact that `value` is always set when a route was found?

For example:

```ts
export interface ResolvedNode<T> {
   node: Node<T> & {value: T}; // the value is always set
   // tslint:disable-next-line:no-any
   params?: {[name: string]: any};
 }
```